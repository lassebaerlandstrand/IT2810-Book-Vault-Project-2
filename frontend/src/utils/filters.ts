import { SortBy, SortOrder } from '@/generated/graphql';

export const DEFAULT_FILTERS = {
  sortBy: SortBy.BookName,
  sortOrder: SortOrder.Asc,
  selectedGenres: [],
  selectedPublishers: [],
  selectedAuthors: [],
  selectedBeforeDate: new Date(2020, 11, 31),
  selectedAfterDate: new Date(1900, 0, 1),
  selectedMinPages: undefined,
  selectedMaxPages: undefined,
  selectedMinRating: undefined,
};

export const getFilterParams = (searchParams: URLSearchParams) => {
  const sortBy = (searchParams.get('sortBy') as SortBy) ?? DEFAULT_FILTERS.sortBy;
  const sortOrder = (searchParams.get('sortOrder') as SortOrder) ?? DEFAULT_FILTERS.sortOrder;
  const authors = searchParams.getAll('authors') ?? DEFAULT_FILTERS.selectedAuthors;
  const publishers = searchParams.getAll('publishers') ?? DEFAULT_FILTERS.selectedPublishers;
  const genres = searchParams.getAll('genres') ?? DEFAULT_FILTERS.selectedGenres;
  const beforeDate = searchParams.get('beforeDate')
    ? new Date(searchParams.get('beforeDate') ?? '')
    : DEFAULT_FILTERS.selectedBeforeDate;
  const afterDate = searchParams.get('afterDate')
    ? new Date(searchParams.get('afterDate') ?? '')
    : DEFAULT_FILTERS.selectedAfterDate;
  const minPages =
    parseInt(searchParams.get('minPages') ?? '', 10) || DEFAULT_FILTERS.selectedMinPages;
  const maxPages =
    parseInt(searchParams.get('maxPages') ?? '', 10) || DEFAULT_FILTERS.selectedMaxPages;
  const minRating =
    parseInt(searchParams.get('minRating') ?? '', 10) || DEFAULT_FILTERS.selectedMinRating;

  return {
    sortBy,
    sortOrder,
    authors,
    publishers,
    genres,
    beforeDate,
    afterDate,
    minPages,
    maxPages,
    minRating,
  };
};
