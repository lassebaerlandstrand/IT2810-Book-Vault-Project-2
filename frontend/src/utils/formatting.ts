/**
 * Formats a number by adding spaces as thousand separators.
 *
 * @param {number} number - The input number to format
 * @returns {string} The formatted number string with space separators
 *
 * @example
 * // returns '1 000'
 * formatNumberWithSpaces(1000);
 *
 * @example
 * // returns '1 000 000'
 * formatNumberWithSpaces(1000000);
 */
export const formatNumberWithSpaces = (number: number) =>
  number.toString().replace(/\B(?=(\d{3})+(?!\d))/g, ' ');

// The regular expression used in the replace method:
// \B - matches a non-word boundary to avoid adding spaces at the start of the string.
// (?=(\d{3})+(?!\d)) - positive lookahead to find positions followed by a group of three digits not followed by another digit.
