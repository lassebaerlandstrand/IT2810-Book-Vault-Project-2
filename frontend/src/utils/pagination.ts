const DEFAULT_PAGE = 1;
const DEFAULT_LIMIT = 25;
const LIMIT_OPTIONS = ['10', '25', '50', '100'];

export const getPaginationParams = (searchParams: URLSearchParams) => {
  const page = Number(searchParams.get('page')) || DEFAULT_PAGE;
  const limit = Number(searchParams.get('limit')) || DEFAULT_LIMIT;

  return { page, limit, DEFAULT_PAGE, DEFAULT_LIMIT, LIMIT_OPTIONS };
};
