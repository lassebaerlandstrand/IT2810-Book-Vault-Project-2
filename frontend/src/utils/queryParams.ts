import { SetURLSearchParams } from 'react-router-dom';

/**
 * Updates the URL search parameters based on the provided key and value.
 * Handles different types of values (single string, array of strings, or undefined).
 * This function handles consecutive calls.
 *
 * @param {SetURLSearchParams} setSearchParams - Function to update the URL search parameters
 * @param {string} key - The key of the query parameter to update
 * @param {string | string[] | undefined} value - The value to set for the query parameter
 *
 * @example
 * updateQueryParams(setSearchParams, 'search', 'newSearchQuery')
 * updateQueryParams(setSearchParams, 'genres', ['tag1', 'tag2'])
 * updateQueryParams(setSearchParams, 'page', '5')
 * updateQueryParams(setSearchParams, 'page', undefined) // Removes the key
 *
 */
export const updateQueryParams = (
  setSearchParams: SetURLSearchParams,
  key: string,
  value: string | string[] | undefined
) => {
  setSearchParams((params) => {
    if (Array.isArray(value)) {
      params.delete(key);
      value.forEach((v) => params.append(key, v));
    } else if (value === undefined || value === '') {
      params.delete(key);
    } else {
      params.set(key, value);
    }
    return params;
  });
};
