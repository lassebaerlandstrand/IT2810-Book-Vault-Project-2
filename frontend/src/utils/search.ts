/**
 * Extracts the search parameter from the URL.
 * This function ensures that the search parameters are correctly parsed and returned in a consistent format.
 *
 * @param {URLSearchParams} searchParams - URL search parameters containing search criteria
 * @returns {Object} Search parameters
 *
 */
export const getSearchParams = (searchParams: URLSearchParams) => {
  const searchValue = searchParams.get('search') || '';
  return { searchValue };
};
