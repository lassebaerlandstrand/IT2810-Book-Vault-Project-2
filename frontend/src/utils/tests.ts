/**
 * Removes random attributes generated by Mantine from DOM elements.
 * This is necessary to ensure consistent snapshots in tests, as Mantine uses random IDs.
 */
export const removeMantineRandomAttributes = () => {
  const attributesToRemove = [
    ...document.body.querySelectorAll('div [id^="mantine"]'),
    ...document.body.querySelectorAll('div [for^="mantine"]'),
  ];
  attributesToRemove.forEach((element) => {
    element.removeAttribute('for');
    element.removeAttribute('id');
    element.removeAttribute('class');
    element.removeAttribute('aria-describedby');
  });
};
