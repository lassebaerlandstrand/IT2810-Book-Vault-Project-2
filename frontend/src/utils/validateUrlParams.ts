import { SortBy, SortOrder } from '@/generated/graphql';
import { getFilterParams } from './filters';
import { getPaginationParams } from './pagination';

const isValidFilters = (sortBy: SortBy, sortOrder: SortOrder): boolean => {
  if (!Object.values(SortBy).includes(sortBy)) {
    return false;
  }

  if (!Object.values(SortOrder).includes(sortOrder)) {
    return false;
  }

  return true;
};

const isValidPagination = (page: number, limit: number, LIMIT_OPTIONS: string[]): boolean => {
  if (isNaN(page) || page <= 0) {
    return false;
  }

  if (isNaN(limit) || !LIMIT_OPTIONS.includes(limit.toString())) {
    return false;
  }

  return true;
};

/**
 * Validates URL search parameters for filters and pagination.
 * Ensures that the parameters are within expected ranges and formats.
 *
 * @param {URLSearchParams} searchParams - URL search parameters containing filter and pagination criteria
 * @returns {boolean} True if all parameters are valid, false otherwise
 */
export const isValidUrlParams = (searchParams: URLSearchParams): boolean => {
  const { sortBy, sortOrder } = getFilterParams(searchParams);
  if (!isValidFilters(sortBy, sortOrder)) {
    return false;
  }

  const { page, limit, LIMIT_OPTIONS } = getPaginationParams(searchParams);
  if (!isValidPagination(page, limit, LIMIT_OPTIONS)) {
    return false;
  }

  return true;
};
