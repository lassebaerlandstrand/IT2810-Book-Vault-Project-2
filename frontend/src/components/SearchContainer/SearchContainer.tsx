import { useState } from 'react';
import { IconArrowRight, IconSearch } from '@tabler/icons-react';
import { useSearchParams } from 'react-router-dom';
import { ActionIcon, Flex, TextInput } from '@mantine/core';
import { getHotkeyHandler } from '@mantine/hooks';
import { DEFAULT_PAGE } from '@/utils/pagination';
import { updateQueryParams } from '@/utils/queryParams';
import { getSearchParams } from '@/utils/search';

const SearchContainer = () => {
  const [searchParams, setSearchParams] = useSearchParams();
  const { searchValue: initialSearchValue } = getSearchParams(searchParams);
  const [searchValue, setSearchValue] = useState(initialSearchValue);

  const performSearch = (searchValue: string) => {
    updateQueryParams(setSearchParams, 'page', DEFAULT_PAGE.toString());
    updateQueryParams(setSearchParams, 'search', searchValue);
  };

  return (
    <Flex justify="center" my={10}>
      <TextInput
        placeholder="Search for books"
        leftSection={<IconSearch />}
        value={searchValue}
        onInput={(event) => {
          setSearchValue((event.target as HTMLInputElement).value);
        }}
        onKeyDown={getHotkeyHandler([['Enter', () => performSearch(searchValue)]])}
        rightSection={
          <ActionIcon onClick={() => performSearch(searchValue)} variant="filled">
            <IconArrowRight size="80%" />
          </ActionIcon>
        }
      />
    </Flex>
  );
};

export default SearchContainer;
