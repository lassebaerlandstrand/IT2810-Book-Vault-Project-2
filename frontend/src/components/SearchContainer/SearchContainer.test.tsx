import { fireEvent, render, screen } from '@test-utils';
import { useSearchParams } from 'react-router-dom';
import { updateQueryParams } from '@/utils/queryParams';
import { getSearchParams } from '@/utils/search';
import SearchContainer from './SearchContainer';

vi.mock('react-router-dom', () => ({
  useSearchParams: vi.fn(),
}));

vi.mock('@/utils/queryParams', () => ({
  updateQueryParams: vi.fn(),
}));

vi.mock('@/utils/search', () => ({
  getSearchParams: vi.fn(),
}));

describe('SearchContainer', () => {
  const mockSetSearchParams = vi.fn();

  beforeEach(() => {
    (useSearchParams as jest.Mock).mockImplementation(() => [
      new URLSearchParams(), // Simulating an empty searchParams initially
      mockSetSearchParams, // This is a mock setter function
    ]);

    (getSearchParams as jest.Mock).mockReturnValue({ searchValue: '' });
  });

  afterEach(() => {
    vi.clearAllMocks();
  });

  it('renders the search input field', () => {
    render(<SearchContainer />);

    const searchInput = screen.getByPlaceholderText('Search for books');
    expect(searchInput).toBeInTheDocument();
  });

  it('updates search value on input change', () => {
    render(<SearchContainer />);

    const inputElement = screen.getByPlaceholderText('Search for books');
    fireEvent.input(inputElement, { target: { value: 'Harry Potter' } });

    expect(inputElement).toHaveValue('Harry Potter');
  });

  it('calls performSearch on Enter key press', async () => {
    render(<SearchContainer />);

    const inputElement = screen.getByPlaceholderText('Search for books');
    fireEvent.input(inputElement, { target: { value: 'Harry Potter' } });
    fireEvent.keyDown(inputElement, { key: 'Enter' });

    expect(updateQueryParams).toHaveBeenCalledWith(mockSetSearchParams, 'page', '1');
    expect(updateQueryParams).toHaveBeenCalledWith(mockSetSearchParams, 'search', 'Harry Potter');
  });

  it('calls performSearch on search button click', async () => {
    render(<SearchContainer />);

    const inputElement = screen.getByPlaceholderText('Search for books');
    fireEvent.input(inputElement, { target: { value: 'Harry Potter' } });
    const searchButton = screen.getByRole('button');

    fireEvent.click(searchButton);

    expect(updateQueryParams).toHaveBeenCalledWith(mockSetSearchParams, 'page', '1');
    expect(updateQueryParams).toHaveBeenCalledWith(mockSetSearchParams, 'search', 'Harry Potter');
  });

  it('matches snapshot', () => {
    const { asFragment } = render(<SearchContainer />);
    const attributesToRemove = document.body.querySelectorAll('div [id^="mantine"]'); // Because Mantine uses random ids which causes snapshots to fail
    attributesToRemove.forEach((element) => {
      element.removeAttribute('id');
      element.removeAttribute('aria-describedby');
    });
    expect(asFragment()).toMatchSnapshot();
  });
});
