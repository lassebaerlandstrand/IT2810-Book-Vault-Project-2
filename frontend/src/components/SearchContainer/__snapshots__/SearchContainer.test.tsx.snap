// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders SearchContainer 1`] = `
<DocumentFragment>
  <style
    data-mantine-styles="true"
  >
    :root{--mantine-font-family: Poppins, Roboto, Helvetica, Arial, sans-serif, Apple Color Emoji, Segoe UI Emoji;--mantine-font-family-headings: Poppins, Roboto, Helvetica, Arial, sans-serif, Apple Color Emoji, Segoe UI Emoji;--mantine-breakpoint-xs: 30em;--mantine-breakpoint-md: 64em;--mantine-breakpoint-xl: 90em;--mantine-breakpoint-xxs: 24em;}
  </style>
  <style
    data-mantine-styles="classes"
  >
    @media (max-width: 29.99375em) {.mantine-visible-from-xs {display: none !important;}}@media (min-width: 30em) {.mantine-hidden-from-xs {display: none !important;}}@media (max-width: 47.99375em) {.mantine-visible-from-sm {display: none !important;}}@media (min-width: 48em) {.mantine-hidden-from-sm {display: none !important;}}@media (max-width: 63.99375em) {.mantine-visible-from-md {display: none !important;}}@media (min-width: 64em) {.mantine-hidden-from-md {display: none !important;}}@media (max-width: 74.99375em) {.mantine-visible-from-lg {display: none !important;}}@media (min-width: 75em) {.mantine-hidden-from-lg {display: none !important;}}@media (max-width: 89.99375em) {.mantine-visible-from-xl {display: none !important;}}@media (min-width: 90em) {.mantine-hidden-from-xl {display: none !important;}}@media (max-width: 23.99375em) {.mantine-visible-from-xxs {display: none !important;}}@media (min-width: 24em) {.mantine-hidden-from-xxs {display: none !important;}}
  </style>
  <div
    class="m_8bffd616 mantine-Flex-root __m__-r0"
    style="justify-content: center; margin-block: calc(0.625rem * var(--mantine-scale));"
  >
    <div
      class="m_46b77525 mantine-InputWrapper-root mantine-TextInput-root"
    >
      <div
        class="m_6c018570 mantine-Input-wrapper mantine-TextInput-wrapper"
        data-variant="default"
        data-with-left-section="true"
        data-with-right-section="true"
      >
        <div
          class="m_82577fc2 mantine-Input-section mantine-TextInput-section"
          data-position="left"
        >
          <svg
            class="tabler-icon tabler-icon-search "
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M10 10m-7 0a7 7 0 1 0 14 0a7 7 0 1 0 -14 0"
            />
            <path
              d="M21 21l-6 -6"
            />
          </svg>
        </div>
        <input
          aria-invalid="false"
          class="m_8fb7ebe7 mantine-Input-input mantine-TextInput-input"
          data-variant="default"
          id="mantine-hsesk81dp"
          placeholder="Search for books"
          value=""
        />
        <div
          class="m_82577fc2 mantine-Input-section mantine-TextInput-section"
          data-position="right"
        >
          <button
            class="mantine-focus-auto mantine-active m_8d3f4000 mantine-ActionIcon-root m_87cf2631 mantine-UnstyledButton-root"
            data-variant="filled"
            style="--ai-bg: var(--mantine-color-gray-filled); --ai-hover: var(--mantine-color-gray-filled-hover); --ai-color: var(--mantine-color-white); --ai-bd: calc(0.0625rem * var(--mantine-scale)) solid transparent;"
            type="button"
          >
            <span
              class="m_8d3afb97 mantine-ActionIcon-icon"
            >
              <svg
                class="tabler-icon tabler-icon-arrow-right "
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M5 12l14 0"
                />
                <path
                  d="M13 18l6 -6"
                />
                <path
                  d="M13 6l6 6"
                />
              </svg>
            </span>
          </button>
        </div>
      </div>
    </div>
    <button
      class="mantine-focus-auto mantine-active m_8d3f4000 mantine-ActionIcon-root m_87cf2631 mantine-UnstyledButton-root"
      data-size="input-sm"
      style="--ai-size: var(--ai-size-input-sm); --ai-bg: var(--mantine-color-gray-filled); --ai-hover: var(--mantine-color-gray-filled-hover); --ai-color: var(--mantine-color-white); --ai-bd: calc(0.0625rem * var(--mantine-scale)) solid transparent;"
      type="button"
    >
      <span
        class="m_8d3afb97 mantine-ActionIcon-icon"
      >
        <svg
          class="tabler-icon tabler-icon-adjustments "
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M4 10a2 2 0 1 0 4 0a2 2 0 0 0 -4 0"
          />
          <path
            d="M6 4v4"
          />
          <path
            d="M6 12v8"
          />
          <path
            d="M10 16a2 2 0 1 0 4 0a2 2 0 0 0 -4 0"
          />
          <path
            d="M12 4v10"
          />
          <path
            d="M12 18v2"
          />
          <path
            d="M16 7a2 2 0 1 0 4 0a2 2 0 0 0 -4 0"
          />
          <path
            d="M18 4v1"
          />
          <path
            d="M18 9v11"
          />
        </svg>
      </span>
    </button>
  </div>
</DocumentFragment>
`;
